core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java

```diff
@@ -245,21 +244,24 @@ public static RelDataType toSql(final RelDataTypeFactory typeFactory,
 
   private static RelDataType toSql(final RelDataTypeFactory typeFactory,
       RelDataType type, boolean mustSetNullability) {
-    RelDataType sqlType = type;
     if (type instanceof RelRecordType) {
       // We do not need to change the nullability of the nested fields,
       // since it can be overridden by the existing implementation of createTypeWithNullability
       // when we treat the nullability of the root struct type.
-      sqlType = typeFactory.createStructType(
-              Lists.transform(type.getFieldList(),
-                field -> toSql(typeFactory, field.getType(), false)),
-              type.getFieldNames());
+      final RelDataType structured = typeFactory.createStructType(
+          type.getFieldList()
+              .stream()
+              .map(field -> toSql(typeFactory, field.getType(), false))
+              .collect(Collectors.toList()),
+          type.getFieldNames());
+      return mustSetNullability
+                 ? typeFactory.createTypeWithNullability(structured, type.isNullable())
```

Comment:
Why we must have a flag `mustSetNullability` ?

Reply:
@danny0405 , possibly we will remove it. This is currently under discussion on the Jira ticket

Reply:
Removed according to discussion in JIRA.